src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+        let federation_version = lazily_resolved_supergraph_config
+            .federation_version()
+            .clone()
+            .unwrap_or(FederationVersion::LatestFedTwo);
+
+        // TODO: Let the supergraph binary exist inside its own task that can respond to being re-installed etc.
+        let supergraph_binary =
+            InstallSupergraph::new(federation_version.clone(), client_config.clone())
+                .install(
+                    None,
+                    lsp_opts.plugin_opts.elv2_license_accepter,
+                    lsp_opts.plugin_opts.skip_update,
+                )
+                .await?;
+
+        // Spin up Runner
+        let mut stream = Runner::default()
+            .setup_subgraph_watchers(
+                lazily_resolved_supergraph_config.subgraphs().clone(),
+                &lsp_opts.plugin_opts.profile,
+                &client_config,
+                500,
+            )
+            .setup_supergraph_config_watcher(lazily_resolved_supergraph_config.clone())
+            .setup_composition_watcher(
+                lazily_resolved_supergraph_config
+                    .fully_resolve_subgraphs(&client_config, &studio_client)
+                    .await
+                    .map_err(ResolveSupergraphConfigError::ResolveSubgraphs)?,
+                supergraph_binary,
+                TokioCommand::default(),
+                FsReadFile::default(),
+                FsWriteFile::default(),
+                OutputTarget::Stdout,
+                Utf8PathBuf::try_from(temp_dir())?,
+            )
+            .run();
+
+        let supergraph_yaml_url =
+            tower_lsp::lsp_types::Url::from_file_path(supergraph_yaml_path)
+                .map_err(|_| anyhow!("Failed to convert supergraph yaml path to url"))?;
+
+        while let Some(event) = stream.next().await {
+            match event {
+                CompositionEvent::Started => {
+                    language_server.composition_did_start().await;
+                }
+                CompositionEvent::Success(CompositionSuccess {
+                    supergraph_sdl,
+                    federation_version,
+                    hints,
+                }) => {
+                    debug!("Successfully composed with version {}", federation_version);
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            Some(supergraph_sdl),
+                            hints.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await;
+                }
+                CompositionEvent::Error(CompositionError::Build { source: errors }) => {
```

Comment:
This is slightly different to the original version because the `dev` refactor does not distinguish at the top level between "Completing with errors" and "Erroring". So we have to pattern match on the error itself to get what we want.